sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/ReturnInFinallyCheck.kt

```diff
@@ -0,0 +1,119 @@
+/*
+ * SonarSource Kotlin
+ * Copyright (C) 2018-2021 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarsource.kotlin.checks
+
+import org.jetbrains.kotlin.psi.KtBreakExpression
+import org.jetbrains.kotlin.psi.KtContinueExpression
+import org.jetbrains.kotlin.psi.KtElement
+import org.jetbrains.kotlin.psi.KtExpressionWithLabel
+import org.jetbrains.kotlin.psi.KtFinallySection
+import org.jetbrains.kotlin.psi.KtLabeledExpression
+import org.jetbrains.kotlin.psi.KtLambdaExpression
+import org.jetbrains.kotlin.psi.KtLoopExpression
+import org.jetbrains.kotlin.psi.KtReturnExpression
+import org.jetbrains.kotlin.psi.KtThrowExpression
+import org.jetbrains.kotlin.psi.KtTreeVisitorVoid
+import org.sonar.check.Rule
+import org.sonarsource.kotlin.api.AbstractCheck
+import org.sonarsource.kotlin.plugin.KotlinFileContext
+
+private class FinallyBlockVisitor(private val report: (KtElement) -> Unit) : KtTreeVisitorVoid() {
+    private var loopDepthCounter = 0
+    private var lambdaDepthCounter = 0
+    private val labelsInFinallyBlock = ArrayList<String>()
+    private val jumpsToLabel = ArrayList<KtExpressionWithLabel>()
+
+    override fun visitFinallySection(finallySection: KtFinallySection) {
+        super.visitFinallySection(finallySection)
+        jumpsToLabel
+            .filter { isEscapingTheBlock(it, labelsInFinallyBlock) }
+            .forEach(report)
+    }
```

Comment:
You can have `try()/finally()` inside the `finally()` block. And this code will report an issue twice.

To my mind, you don't need to wait till the end of finally block, but report directly on labeled expressions, because you cannot escape the label, that is out of your scope. In that case you don't need `jumpsToLabel` property at all